### PR TITLE
First loss implementation for dualTC

### PR DESCRIPTION
**Pull request by linesd**
New loss function for dualTC + changes in main.

### CODE REVIEW

File: disvae/models/losses.py [hunk 1]
```diff
@@ -359,10 +365,10 @@ def __call__(self, data, recon_batch, latent_dist, is_train, storer,
         rec_loss = _reconstruction_loss(data, recon_batch,
                                         storer=storer,
                                         distribution=self.rec_dist)
-        log_pz, log_qz, log_prod_qzi, log_q_zCx = _get_log_pz_qz_prodzi_qzCx(latent_sample,
-                                                                             latent_dist,
-                                                                             self.n_data,
-                                                                             is_mss=self.is_mss)
+        log_pz, log_qz, log_prod_qzi, log_q_zCx, _ = _get_log_pz_qz_prodzi_qzCx_qz_knc(latent_sample,
```
Comment (YannDubs) [inline disvae/models/losses.py]: should be `_get_log_pz_qz_prodzi_qzCx_qzknc` as underscores separate each outputs

File: disvae/models/losses.py [hunk 2]
```diff
@@ -537,6 +616,15 @@ def _get_log_pz_qz_prodzi_qzCx(latent_sample, latent_dist, n_data, is_mss=True):
...
+    log_prod_qzi = torch.logsumexp(mat_log_qz, dim=1, keepdim=False).sum(dim=1)
+
+    knc = _sum_along_dim_exclude_self(mat_log_qz, dim=2)
+    log_qz_knc = torch.logsumexp(knc, dim=1, keepdim=False).sum(dim=1)
+
+    return log_pz, log_qz, log_prod_qzi, log_q_zCx, log_qz_knc
 
-    return log_pz, log_qz, log_prod_qzi, log_q_zCx
+def _sum_along_dim_exclude_self(tensor_in, dim):
+    dims_list = list(tensor_in.shape)
```
Comment (YannDubs) [inline disvae/models/losses.py]: no need, you can enumerate on the tuple returned by shape

File: disvae/models/losses.py [hunk 3]
```diff
@@ -537,6 +616,15 @@ def _get_log_pz_qz_prodzi_qzCx(latent_sample, latent_dist, n_data, is_mss=True):
...
-    log_prod_qzi = torch.logsumexp(mat_log_qz, dim=1, keepdim=False).sum(1)
+    log_prod_qzi = torch.logsumexp(mat_log_qz, dim=1, keepdim=False).sum(dim=1)
+
+    knc = _sum_along_dim_exclude_self(mat_log_qz, dim=2)
+    log_qz_knc = torch.logsumexp(knc, dim=1, keepdim=False).sum(dim=1)
+
+    return log_pz, log_qz, log_prod_qzi, log_q_zCx, log_qz_knc
 
-    return log_pz, log_qz, log_prod_qzi, log_q_zCx
+def _sum_along_dim_exclude_self(tensor_in, dim):
```
Comment (YannDubs) [inline disvae/models/losses.py]: nice implementation, should be documented, remove the leading underscore and moved to `math.py` that's the kind of function we might use again not only in losses

**Comment by YannDubs** [comment]
Minor comments, nothing that makes it bad to test so I'll test it now
